### PR TITLE
alt highlighting

### PR DESCRIPTION
**Pull request by qndel**
Rewrite of https://github.com/diasurgical/devilutionX/pull/1345

### CODE REVIEW

File: Source/qol/itemlabels.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+{
+	if (altPressed == invertHighlightToggle)
+		return;
+	ItemStruct *it = &items[id];
+
+	char textOnGround[64];
+	if (it->_itype == ITYPE_GOLD)
+		std::sprintf(textOnGround, "%i gold", it->_ivalue);
+	else
+		std::strcpy(textOnGround, it->_iIdentified ? it->_iIName : it->_iName);
```
Comment (AJenbo) [inline Source/qol/itemlabels.cpp:63]: You can avoid the string copy:
```suggestion
	const char *textOnGround;
	if (it->_itype == ITYPE_GOLD) {
		std::sprintf(tempstr, "%i gold", it->_ivalue);
		textOnGround = tempstr;
	} else {
		textOnGround = it->_iIdentified ? it->_iIName : it->_iName;
	}
```

File: Source/qol/itemlabels.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		labelCenterOffsets[index].emplace((itemBounds.first + itemBounds.second) / 2);
+	}
+
+	x += *labelCenterOffsets[index];
+	y -= TILE_HEIGHT;
+	if (!zoomflag) {
+		x <<= 1;
+		y <<= 1;
+	}
+	x -= nameWidth >> 1;
```
Comment (AJenbo) [inline Source/qol/itemlabels.cpp:78]: The compiler can do optimizations for us.
```suggestion
	if (!zoomflag) {
		x *= 2;
		y *= 2;
	}
	x -= nameWidth / 2;
```

File: Source/qol/itemlabels.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		return false;
+	if ((chrflag || questlog) && MouseX < SPANEL_WIDTH && MouseY <= SPANEL_HEIGHT)
+		return false;
+	if (MouseY >= PANEL_TOP && MouseX >= PANEL_LEFT && MouseX <= PANEL_LEFT + PANEL_WIDTH)
+		return false;
+
+	return true;
+}
+
+void FastDrawHorizLine(const CelOutputBuffer &out, int x, int y, int width, Uint8 col)
```
Comment (AJenbo) [inline Source/qol/itemlabels.cpp:94]: This is identical to `UnsafeDrawHorizontalLine`, but probably better to use `DrawHorizontalLine` as you do not appear to be doing any bound checks.

**Comment by Chance4us** [comment]
By the way, thank you much for looking at it again. I certainly wouldn't have made it so quickly or not at all.

**Comment by Chance4us** [comment]
@AJenbo I accidentally opened some weird Review window. I hope it's not a problem.

File: Source/engine/render/cel_render.cpp [hunk 4]
```diff
@@ -713,7 +713,7 @@ void CelBlitOutlineTo(const CelOutputBuffer &out, uint8_t col, int sx, int sy, c
 std::pair<int, int> MeasureSolidHorizontalBounds(const CelSprite &cel, int frame)
 {
 	int nDataSize;
-	auto src = reinterpret_cast<const uint8_t *>(CelGetFrame(cel.Data(), frame, &nDataSize));
+	auto src = reinterpret_cast<const uint8_t *>(CelGetFrameClipped(cel.Data(), frame, &nDataSize));
```
Comment (glebm) [inline Source/engine/render/cel_render.cpp:716]:
```suggestion
	const byte *src = CelGetFrameClipped(cel.Data(), frame, &nDataSize);
```

**Comment by AJenbo** [comment]
> @AJenbo I accidentally opened some weird Review window. I hope it's not a problem.

evidently you didn't submit a review

**Review by Chance4us — CHANGES_REQUESTED** [review]

File: Source/qol/itemlabels.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		if (MouseX >= label.pos.x && MouseX <= label.pos.x + label.width && MouseY >= label.pos.y - height && MouseY <= label.pos.y) {
+			if (!gmenu_is_active() && PauseMode == 0 && !deathflag && IsMouseOverGameArea()) {
+				isLabelHighlighted = true;
+				cursmx = itm.position.x;
+				cursmy = itm.position.y;
+				pcursitem = label.id;
+			}
+		}
+		if (pcursitem == label.id)
+			FillRect(out, label.pos.x, label.pos.y - height, label.width + 1, height, PAL8_BLUE + 6);
```
[the remark this answers is not in the document]
Comment (Chance4us) [reply Source/qol/itemlabels.cpp:156]: I suggest the following change for "clipping".
		if ((label.pos.x + label.width < gnScreenWidth) &&
			(label.pos.y + height < gnScreenHeight) &&
			(label.pos.y - height - height > 0)) {
		if (pcursitem == label.id)
			FillRect(out, label.pos.x, label.pos.y - height, label.width + 1, height, PAL8_BLUE + 6);
		else
			DrawHalfTransparentRectTo(out, label.pos.x, label.pos.y - height, label.width + 1, height);
		DrawString(out, label.text.c_str(), { label.pos.x, label.pos.y, label.width, height }, itm.getTextColor());
		}
These change prevent the crash, which does not happen every time, but still causes issue.

In addition, the appearance improves as follows. 
Without clipping:
[noclipping.mkv.zip](https://github.com/diasurgical/devilutionX/files/6541801/noclipping.mkv.zip)

With clipping:
[clipping.mkv.zip](https://github.com/diasurgical/devilutionX/files/6541804/clipping.mkv.zip)

Savegame for testing purposes:
[single_0.hsv.zip](https://github.com/diasurgical/devilutionX/files/6541809/single_0.hsv.zip)

[the remark this answers is not in the document]
Comment (qndel) [reply Source/qol/itemlabels.cpp:156]: dunno, they are a mess

**Comment by Chance4us** [comment]
@qndel I can not use this github stuff so well. You can see my suggestions by clicking "Show resolved" to see the text.
Thanks in advance.

File: Source/qol/itemlabels.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Chance4us) [reply Source/qol/itemlabels.cpp:156]: Do you think the clipping is okay?

File: Source/qol/itemlabels.cpp (hunk 3 above)
Comment (glebm) [reply Source/qol/itemlabels.cpp:96]: @qndel You marked this resolved but the code is still the same. Forgot to push?

**Comment by AJenbo** [comment]
> @qndel I can not use this github stuff so well. You can see my suggestions by clicking "Show resolved" to see the text.
> Thanks in advance.

I think he knows, considering that he was the one to mark it as resolved to avoid a lot of message being shown.

File: Source/qol/itemlabels.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (qndel) [reply Source/qol/itemlabels.cpp:156]: seems clipping was incorrect in general

File: Source/qol/itemlabels.cpp (hunk 3 above)
Comment (qndel) [reply Source/qol/itemlabels.cpp]: no, was poking :P

**Comment by Chance4us** [comment]
Wow, all respect. That feels great now @qndel.
Now we can merge.

**Review by glebm — APPROVED** [review]

File: Source/qol/itemlabels.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
+/**
+* @file itemlabels.cpp
+*
+* Adds item labels QoL feature
+*/
+
```
Comment (AJenbo) [inline Source/qol/itemlabels.cpp:6]: We only do this for header files now.

```suggestion
```

File: Source/qol/itemlabels.cpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	const char *textOnGround;
+	if (it->_itype == ITYPE_GOLD) {
+		std::sprintf(tempstr, _("%i gold"), it->_ivalue);
+		textOnGround = tempstr;
+	} else {
+		textOnGround = it->_iIdentified ? it->_iIName : it->_iName;
+	}
+
+	int nameWidth = GetLineWidth(textOnGround);
+	BYTE index = ItemCAnimTbl[it->_iCurs];
```
Comment (AJenbo) [inline Source/qol/itemlabels.cpp:74]: Avoid WinAPI types
```suggestion
	uint8_t index = ItemCAnimTbl[it->_iCurs];
```

**Review by AJenbo — DISMISSED** [review]

File: Source/qol/itemlabels.cpp (hunk 7 above)
Comment (AJenbo) [reply Source/qol/itemlabels.cpp:68]: Actually doesn't look like there is a reason to have a specific size here.
```suggestion
	int index = ItemCAnimTbl[it->_iCurs];
```

**Comment by Chance4us** [comment]
@qndel I've just tested a little more extensively and for some reason the item label color in dungeon is always red.
But in town the color is perfect.

**Pull request merged**